### PR TITLE
Feature/csv data preview

### LINKED ISSUE #6439: [FR] Format CSV output as a table in ArtifactViewer
### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

If a logged artifact is a CSV file, we try parsing the file and laying out a slice of the data in an embedded table in the artifact viewer.

### Motivation

> #### What is the use case for this feature?
To have a more aesthetically pleasing experience viewing tabular data in MLflow

> #### Why is this use case valuable to support for MLflow users in general?
As users rely more on the artifact viewer as a part of their mlops experience, this could create less time switching to another app to show a slice of training data or testing results.

> #### Why is this use case valuable to support for your project(s) or organization?
We don't shy away from showing our business partners our experiment runs. Being able to preview CSV files (especially training data) would be a big help when we present our work.

> #### Why is it currently difficult to achieve this use case?
* You would need to utilize a CSV parser dependency to be able to lay out the data preview

### Details

* If a CSV file cannot be parsed, fallback to the plaintext preview of the file with a banner saying it was unable to be parsed
* To limit performance issues in the browser, maybe we limit the CSV preview to the first `N` rows
* We can reuse the CSS in the experiments table in the experiments page
* Utilize an alternating cell background color to ease the eyes
* Important that we don't turn this feature into a tabular data viewer and gently encourage the user to download their CSV file to view in their spreadsheet editor of choice by:
1. Not allowing a sort feature on the column header, and 
2. Not allowing the user to re-order columns

See below for current way CSVs are visualized in MLflow
Screen Shot 2022-08-10 at 8 02 30 PM

See below on implementations from similar tools

Snowflake

snowflake_table_example

Dataiku

dataiku_table_example

GitHub

github

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [X] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by nnethery**
## Related Issues/PRs

Close #6439

## What changes are proposed in this pull request?

Add a table visualization to CSV artifacts in the artifacts viewer in a specific experiment's run page.

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

A script to log data for testing:

```python
import tempfile
import requests
import zipfile
import pandas as pd
import mlflow
from pathlib import Path


def unzip(path, dst):
    with zipfile.ZipFile(path, "r") as zip_ref:
        zip_ref.extractall(dst)


def download_url(url, save_path, chunk_size=128):
    r = requests.get(url, stream=True)
    with open(save_path, "wb") as fd:
        for chunk in r.iter_content(chunk_size=chunk_size):
            fd.write(chunk)


with tempfile.TemporaryDirectory() as tmpdir:
    tmpdir = Path(tmpdir)
    url = "https://github.com/mlflow/mlflow/files/9454188/Airbnb_Open_Data.csv.zip"
    name = url.split("/")[-1]
    zip_path = tmpdir.joinpath(name)
    download_url(url, zip_path)
    dst = tmpdir.joinpath("unzipped")
    unzip(zip_path, dst)
    csv_file = next(dst.glob("*.csv"))
    print(pd.read_csv(csv_file).head(1000))

    with mlflow.start_run():
        mlflow.log_artifact(csv_file)
```

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

CSV file artifacts that are logged in experiments are now visualized with a table in the artifact viewer. This table visualizes a slice of the data up to the first 500 rows and the slice is sortable via the column headers.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by nnethery** [comment]
Hello maintainers, I would really like some help on this pull request on the following aspects:

- The header row and its column separators jitter when the data is loaded into the grid
- I would really like the header row to be sticky as is the normal convention of `ag-grid` but I can't figure it out. I suspect when I set `overflow: scroll` on the grid CSS overrides it is messing something up.
- When the number of columns is less than the viewport, it does not fill the entire width of the viewport (test this with a 3 or 4 column CSV with only integers as the row values)

**Comment by nnethery** [comment]
I see that tab separated value files are also still common so I will add support for this in the next commit

**Comment by dbczumar** [comment]
> Hello maintainers, I would really like some help on this pull request on the following aspects:
> 
> * The header row and its column separators jitter when the data is loaded into the grid
> * I would really like the header row to be sticky as is the normal convention of `ag-grid` but I can't figure it out. I suspect when I set `overflow: scroll` on the grid CSS overrides it is messing something up.
> * When the number of columns is less than the viewport, it does not fill the entire width of the viewport (test this with a 3 or 4 column CSV with only integers as the row values)
> * The redrawing of data in the grid on a window resize is very slow

@harupy @sunishsheth2009 Can you review this PR and help to address these points from @nnethery?

**Comment by harupy** [comment]
@nnethery Is there a reason we need to use `AgGrid`? Can we use the `Table` component in `@databricks/design-system`? Here's a quick demo using the `Table` component:


https://user-images.githubusercontent.com/17039389/187144219-b28ac13d-0526-4543-94a7-7571631e74ef.mov

**Comment by nnethery** [comment]
@harupy I tried a bigger CSV that has long strings as values and column headers where the names are longer than the values they contain. I attached it for your own testing and you can pull the latest commit to try. I apologize that I don't know the style overrides available in the custom `Table` component. Any suggestions?

Screen Shot 2022-08-30 at 3 54 01 PM
Screen Shot 2022-08-30 at 4 04 01 PM
[Airbnb_Open_Data.csv.zip](https://github.com/mlflow/mlflow/files/9454188/Airbnb_Open_Data.csv.zip)

**Comment by harupy** [comment]
@nnethery Setting the column props might help: https://ant.design/components/table/#Column.

**Comment by nnethery** [comment]
> @nnethery Setting the column props might help: https://ant.design/components/table/#Column.
> 
> `databricks/design-system` was created based on `antd` so (not all) props available in `antd` should be available in `databricks/design-system` as well.

@harupy Ok I updated the branch. The best solution I could figure out with `antd` is just hard-coding a column width and adding ellipsis if that isn't enough room for the header. It's a little too much space if the header is long but column values are short, but I think it looks good enough to ship.

**Comment by harupy** [comment]
@nnethery Thanks for the udpate! Can you take a screenshot and share it?

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:8]: Can we convert this to a functional component?

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:107]: Is it possible to parse only the first `MAX_ROW_LENGTH` rows?

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:60]: Can we use `FormattedMessage` for internationalization?

Example:

https://github.com/mlflow/mlflow/blob/4c0f21025322f4681e5c4cf01d2ad57e0eb93115/mlflow/server/js/src/experiment-tracking/components/MetricView.js#L28-L35

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:81]:
```suggestion
          <span css={{ display: 'flex', justifyContent: 'center' }}>{rowPreviewMessage}</span>
```

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (nnethery) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js]: This has been resolved by using a `preview` config option in Papaparse which streams each row and aborts the serialization of the rows once the row limit is met. You should notice a performance boost. 

Perhaps for a second iteration of this feature, we can avoid downloading the whole CSV and stream the file’s bytes and parse the rows as we go.

**Comment by nnethery** [comment]
> @nnethery Thanks for the udpate! Can you take a screenshot and share it?

Screen Shot 2022-09-06 at 9 51 22 AM

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:16]:
```suggestion
  const MAX_NUM_ROWS = 500;
```

`MAX_ROW_LENGTH` sounds like the maximum number of elements in a row. What do you think?

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.test.js
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.test.js:22]: Can we test the following behaviors?

1. Render CSV /TSV file fine.
2. When the number of rows is > 500, only show the first 500 rows.
3. Render the raw text if it's not valid CSV.

**Comment by harupy** [comment]
https://user-images.githubusercontent.com/17039389/188789900-748f49b7-5767-4151-b176-400c42307f7d.mov

(for other reviewers)

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactPage.js
Comment (hubertzub-db) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactPage.js:19]: Is it possible to lazily import this component here (using `React.lazy`)? This should end up having `papaparse` dependency in the separate chunk, keeping the main bundle clean and thin ✨

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (harupy) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:81]: Why are we checking whether `a` is NaN here?

Comment (nnethery) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:81]: I add a check if a cell value is not a number and run a string compare function for the sorter. `papaparse` casts parsed data to their most likely data type so I'm just accounting for numbers and strings in the sorter.

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.test.js
Comment (nnethery) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.test.js:28]: I'm having some issues handling test case 2, can you please review the last test in `ShowArtifactTableView.test.js` in the latest commit and help me understand why `wrapper` is an empty array?

Comment (harupy) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.test.js:28]: `mount` needs to be replaced with `mountWithIntl` since `ShowArtifactTableView` contains `FormattedMessage`.

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (nnethery) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js]: After double-checking, you are correct. Everything comes in as a string, and `null`/`undefined` becomes an empty string. I updated the sorter accordingly.

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by harupy** [comment]
@hubertzub-db Could you take another look?

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (sunishsheth2009) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:12]: I am curious if we could use the [RequestStateWrapper](https://github.com/mlflow/mlflow/blob/4c0f21025322f4681e5c4cf01d2ad57e0eb93115/mlflow/server/js/src/common/components/RequestStateWrapper.js) here rather than dealing with loading and error states. I think we can let this component handle that for us. What do you think?

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (hubertzub-db) [inline mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:21]: there's a missing dependency here - please add it or add a eslint ignore directive

**Review by hubertzub-db — APPROVED** [review]
Looks good! Just left a comment re eslint warning and +1 to @sunishsheth2009 's comment: https://github.com/mlflow/mlflow/pull/6567#discussion_r966639800

File: mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js
Comment (nnethery) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:12]: I will take a crack at this before merging!

Comment (nnethery) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:12]: @sunishsheth2009 Is this just for network requests? There are two main errors that could occur with this component: 
1. A call to download the artifact from the artifact server which has its own function wrapped around the API call
2. An error parsing the CSV file in the browser which does not call to the server

Is this wrapper still suitable for these errors and loading?

Comment (sunishsheth2009) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:12]: I think if we want to use [RequestStateWrapper](https://github.com/mlflow/mlflow/blob/4c0f21025322f4681e5c4cf01d2ad57e0eb93115/mlflow/server/js/src/common/components/RequestStateWrapper.js) we need to use redux pattern to use request_ids to work correctly. If we feel that we don't want to use redux here then I think we are fine leaving this approach as it is.
cc @harupy @hubertzub-db thoughts?

Comment (hubertzub-db) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:12]: I think it's fine to leave it as it is for now - looking at the forthcoming plans for improving the API connectivity code and restructuring the data store, there will be an opportunity to unify it anyway.

Comment (nnethery) [reply mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js:12]: Sounds good, all. That’s all the changes left on my end!

**Comment by harupy** [comment]
@nnethery Thanks for the contribution!

@sunishsheth2009 @hubertzub-db Thanks for the review :)

**Pull request merged**
